registry/session.go

```diff
@@ -84,7 +84,13 @@ func (tr *authTransport) RoundTrip(orig *http.Request) (*http.Response, error) {
 	if req.Header.Get("Authorization") == "" {
 		if req.Header.Get("X-Docker-Token") == "true" && len(tr.Username) > 0 {
 			req.SetBasicAuth(tr.Username, tr.Password)
-		} else if len(tr.token) > 0 {
+		} else if len(tr.token) > 0 &&
+			// Authorization should not be set on 302 redirect for untrusted locations.
+			// This logic mirrors the behavior in AddRequiredHeadersToRedirectedRequests.
+			// As the authorization logic is currently implemented in RoundTrip,
+			// a 302 redirect is detected by looking at the Referer header as go http package adds said header.
+			// This is safe as Docker doesn't set Referer in other scenarios.
+			(req.Header.Get("Referer") == "" || trustedLocation(orig)) {
```

Comment:
Ideally yes, the problem is that the http 302 redirect handling is taken care of by the go http client library (https://golang.org/src/net/http/client.go amusingly enough the function starts on line 302) 

In registry.go, there is a similar piece of code that is able to use CheckRedirects for this, as the headers are set earlier auth.go. This won't work in session.go without significant refactoring as the header is set in RoundTrip which is called for each http request, including the recursive calls from the go client code resolving the 302 redirects.

The only way I found to distinguish in RoundTrip between direct requests and 302 redirect requests is by checking the "Referer" header.

There might be a way to refactor the code in session.go to deal with this like in registry.go, but that seemed like a larger work item, I'd be happy to discuss, but my first priority was to get 1.7 unbroken for the unauthenticated scenario on custom registries..